eval/src/vespa/eval/tensor/partial_update.cpp

```diff
@@ -0,0 +1,347 @@
+            lookup_view_dims[i] = i;
+        }
+    }
+    ~SparseCoords();
+};
+SparseCoords::~SparseCoords() = default;
+
+struct AddressHandler {
+    std::vector<DimCase> how;
+    DenseCoords target_coords;
+    SparseCoords for_output;
+    SparseCoords from_modifier;
+    bool valid;
+
+    AddressHandler(const ValueType &input_type,
+                   const ValueType &modifier_type)
+        : how(), target_coords(),
+          for_output(input_type.count_mapped_dimensions()),
+          from_modifier(modifier_type.count_mapped_dimensions()),
+          valid(true)
+    {
+        if (! modifier_type.is_sparse()) {
+            LOG(error, "Unexpected non-sparse modifier tensor, type is %s",
+                modifier_type.to_spec().c_str());
+            valid = false;
+            return;
+        }
+        // analyse dimensions
+        auto visitor = overload {
+            [&](visit_ranges_either, const auto &) { valid = false; },
+            [&](visit_ranges_both, const auto &a, const auto &) {
+                how.push_back(a.is_mapped() ? DimCase::MAPPED_MATCH : DimCase::CONV_TO_INDEXED);
+            }
+        };
+        const auto & input_dims = input_type.dimensions();
+        const auto & modifier_dims = modifier_type.dimensions();
+        visit_ranges(visitor,
+                     input_dims.begin(), input_dims.end(),
+                     modifier_dims.begin(), modifier_dims.end(),
+                     [](const auto &a, const auto &b){ return (a.name < b.name); });
+        if (! valid) {
+            LOG(error, "Value type %s does not match modifier type %s (should have same dimensions)",
+                input_type.to_spec().c_str(),
+                modifier_type.to_spec().c_str());
+            return;
+        }
+        assert(input_dims.size() == modifier_dims.size());
+        assert(input_dims.size() == how.size());
+        for (const auto & dim : input_type.dimensions()) {
+            if (dim.is_indexed()) {
+                target_coords.add_dim(dim.size);
+            }
+        }
+    }
+
+    void handle_address()
+    {
+        target_coords.clear();
+        auto out = for_output.addr.begin();
+        for (size_t i = 0; i < how.size(); ++i) {
+            if (how[i] == DimCase::CONV_TO_INDEXED) {
+                target_coords.with(from_modifier.addr[i]);
+            } else {
+                *out++ = from_modifier.addr[i];
+            }
+        }
+        assert(out == for_output.addr.end());
+        assert(target_coords.dim == target_coords.dim_sizes.size());
+    }
+
+    ~AddressHandler();
+};
+AddressHandler::~AddressHandler() = default;
+
+template <typename CT, typename ICT = CT, typename SkipFun>
+void copy_tensor_with_filter(const Value &input,
+                             size_t dsss,
+                             SparseCoords &addrs,
+                             ValueBuilder<CT> &builder,
+                             SkipFun && skip_subspaces)
```

Comment:
I suggest switching from skip to keep semantic on the predicate. Also, I strongly suggest passing both the address and the subspace to avoid depending too much on side-effect on objects passed into this function.